mods/bucket/init.lua

```diff
@@ -152,6 +157,11 @@ minetest.register_craftitem("bucket:bucket_empty", {
 			minetest.add_node(pointed_thing.under, {name="air"})
 
 			return ItemStack(giving_back)
+		else
+			-- non-liquid nodes will have their on_punch triggered
+			local node_def = minetest.registered_nodes[node.name]
+			node_def.on_punch(pointed_thing.under, node, user, pointed_thing)
```

Comment:
The previous testing for `on_punch` might not have been necessary, but the `nil`-check for `node_def` to deal with punching unknown nodes was. :-)
You can just return the `on_use` passed `itemstack` in that case.